### PR TITLE
Patch LP#1902255

### PR DESCRIPTION
**Pull request by VariableDeclared**
## Description

This patch aims to solve a manual provider issue where manual machine checks will fail even though the target controller is valid.

migrationtarget has been changed to check for manual machines and skip the cloud instances check in this case.

## Checklist

 - [NA] Requires a [pylibjuju](https://github.com/juju/python-libjuju) change
 - [NA ] Added [integration tests](https://github.com/juju/juju/tree/develop/tests) for the PR
 - [NA] Added or updated [doc.go](https://discourse.jujucharms.com/t/readme-in-packages/451) related to packages changed
 - [NA] Comments answer the question of why design decisions were made

## QA steps

*Please replace with how we can verify that the change works.*

```sh
# Create 4 manual machines, setup SSH.


#set -e
export DEBUG_JUJU=1
JUJU=$JUJU_BUILD_DIR
# Controller 0
for machine in node01 node02 node03; do \
        if [ $(maas root machines read | jq -r '.[] | select(.hostname == "'$machine'") | .status_name') = 'Ready' ]
        then
            maas root machine deploy $(maas root machines read | jq -r '.[] | select(.hostname == "'$machine'") | .system_id')
        fi
done

# Controller 1
for machine in node08 node09 node10; do \
        if [ $(maas root machines read | jq -r '.[] | select(.hostname == "'$machine'") | .status_name') = 'Ready' ]
        then
        maas root machine deploy $(maas root machines read | jq -r '.[] | select(.hostname == "'$machine'") | .system_id')
        fi
done
# Controller 0 - model 0
machine=node04
if [ $(maas root machines read | jq -r '.[] | select(.hostname == "'$machine'") | .status_name') = 'Ready' ]
    then
    maas root machine deploy $(maas root machines read | jq -r '.[] | select(.hostname == "'$machine'") | .system_id')
fi
# Controller 1 - model 0
machine=node05
if [ $(maas root machines read | jq -r '.[] | select(.hostname == "'$machine'") | .status_name') = 'Ready' ]
    then
    maas root machine deploy $(maas root machines read | jq -r '.[] | select(.hostname == "'$machine'") | .system_id')
fi

function wait_for_deployed() {
for machine in node01 node02 node03 node04 node05 node08 node09 node10; do
    echo "Machine $(maas root machines read | jq -r '.[] | select(.hostname == "'$machine'") | .status_name')"
    until [ $(maas root machines read | jq -r '.[] | select(.hostname == "'$machine'") | .status_name') = 'Deployed' ];
    do
      echo "Deployment is running......"
      sleep 10s
    done;
done

}

CONTROLLER_0_NODE_0=$(maas root machines read | jq -r '.[] | select(.hostname == "node01") | .ip_addresses[0]')
CONTROLLER_0_NODE_1=$(maas root machines read | jq -r '.[] | select(.hostname == "node02") | .ip_addresses[0]')
CONTROLLER_0_NODE_2=$(maas root machines read | jq -r '.[] | select(.hostname == "node03") | .ip_addresses[0]')

CONTROLLER_1_NODE_0=$(maas root machines read | jq -r '.[] | select(.hostname == "node08") | .ip_addresses[0]')
CONTROLLER_1_NODE_1=$(maas root machines read | jq -r '.[] | select(.hostname == "node09") | .ip_addresses[0]')
CONTROLLER_1_NODE_2=$(maas root machines read | jq -r '.[] | select(.hostname == "node10") | .ip_addresses[0]')
CONTROLLER_0_MODEL_0_NODE_1=$(maas root machines read | jq -r '.[] | select(.hostname == "node04") | .ip_addresses[0]')
CONTROLLER_1_MODEL_0_NODE_1=$(maas root machines read | jq -r '.[] | select(.hostname == "node05") | .ip_addresses[0]')
CONTROLLER_0_CLOUD=test-cloud-controller-0
CONTROLLER_1_CLOUD=test-cloud-controller-1
CONTROLLER_0_CLOUD_CONTROLLER=test-cloud-controller-0-default
CONTROLLER_1_CLOUD_CONTROLLER=test-cloud-controller-1-default
CONTROLLER_0_MODEL_0=model0-$CONTROLLER_0_CLOUD
CONTROLLER_1_MODEL_0=model0-$CONTROLLER_1_CLOUD

cat <<END > $CONTROLLER_0_CLOUD.yaml
clouds:
  $CONTROLLER_0_CLOUD:
    type: manual
    auth-types: [empty]
    endpoint: ubuntu@$CONTROLLER_0_NODE_0 # Endpoint?
    regions:
      default: {}
END

$JUJU remove-cloud $CONTROLLER_0_CLOUD --client
$JUJU add-cloud --client -f $CONTROLLER_0_CLOUD.yaml

$JUJU bootstrap --show-log --logging-config="<root>=TRACE" --build-agent $CONTROLLER_0_CLOUD

$JUJU add-machine -m controller ssh:ubuntu@$CONTROLLER_0_NODE_1
$JUJU add-machine -m controller ssh:ubuntu@$CONTROLLER_0_NODE_2
$JUJU enable-ha --to=1,2


cat <<END > $CONTROLLER_1_CLOUD.yaml
clouds:
  $CONTROLLER_1_CLOUD:
    type: manual
    auth-types: [empty]
    endpoint: ubuntu@$CONTROLLER_1_NODE_0 # Endpoint?
    regions:
      default: {}
END

$JUJU remove-cloud $CONTROLLER_1_CLOUD --client
$JUJU add-cloud --client -f $CONTROLLER_1_CLOUD.yaml

$JUJU bootstrap --show-log --logging-config="<root>=TRACE" --build-agent $CONTROLLER_1_CLOUD

$JUJU add-machine -m controller ssh:ubuntu@$CONTROLLER_1_NODE_1
$JUJU add-machine -m controller ssh:ubuntu@$CONTROLLER_1_NODE_2
$JUJU enable-ha --to=1,2

CONTROLLER_0_MODEL_0=model0-$CONTROLLER_0_CLOUD
CONTROLLER_1_MODEL_0=model0-$CONTROLLER_1_CLOUD
$JUJU add-model -c $CONTROLLER_0_CLOUD_CONTROLLER $CONTROLLER_0_MODEL_0
$JUJU add-model -c $CONTROLLER_1_CLOUD_CONTROLLER $CONTROLLER_1_MODEL_0

$JUJU add-machine -m $CONTROLLER_0_CLOUD_CONTROLLER:$CONTROLLER_0_MODEL_0 ssh:ubuntu@$CONTROLLER_0_MODEL_0_NODE_1
$JUJU add-machine -m $CONTROLLER_1_CLOUD_CONTROLLER:$CONTROLLER_1_MODEL_0 ssh:ubuntu@$CONTROLLER_1_MODEL_0_NODE_1


$JUJU deploy ubuntu --to 0 -m $CONTROLLER_0_CLOUD_CONTROLLER:$CONTROLLER_0_MODEL_0
$JUJU deploy ubuntu --to 0 -m $CONTROLLER_1_CLOUD_CONTROLLER:$CONTROLLER_1_MODEL_0
lly 


juju add-cloud -c $CONTROLLER_0_CLOUD_CONTROLLER -f  $CONTROLLER_1_CLOUD.yaml

juju migrate $CONTROLLER_0_MODEL_0 $CONTROLLER_1_CLOUD_CONTROLLER 

A success will result in model-two appearing on test-cloud-controller-1

```

## Documentation changes

*Please replace with any notes about how it affects current user workflow, CLI, or API.*

## Bug reference

[LP#1902255](https://bugs.launchpad.net/juju/+bug/1902255)

### CODE REVIEW

**Comment by juanmanuel-tirado** [comment]
Please fulfil the PR description.

**Comment by VariableDeclared** [comment]
apologies @juanmanuel-tirado that's done now

**Comment by juanmanuel-tirado** [comment]
Could you please provide some QA steps.

**Comment by VariableDeclared** [comment]
@juanmanuel-tirado added maas based QA steps

**Comment by wallyworld** [comment]
There's some failing manual provider unit tests
```
00:24:27 FAIL: environ_test.go:48: environSuite.TestInstances
00:24:27 
00:24:27 [LOG] 0:00.000 WARNING juju.environs.config unknown config field "ca-cert"
00:24:27 [LOG] 0:00.000 WARNING juju.environs.config unknown config field "ca-private-key"
00:24:27 [LOG] 0:00.000 WARNING juju.environs.config unknown config field "controller-uuid"
00:24:27 [LOG] 0:00.000 WARNING juju.environs.config unknown config field "controller-uuid"
00:24:27 [LOG] 0:00.000 WARNING juju.environs.config unknown config field "ca-cert"
00:24:27 [LOG] 0:00.000 WARNING juju.environs.config unknown config field "ca-private-key"
00:24:27 environ_test.go:57:
00:24:27     c.Assert(err, jc.ErrorIsNil)
00:24:27 ... value *errors.notFound = &errors.unformatter{message:"instances not found", cause:error(nil), previous:error(nil), file:"github.com/juju/juju/environs/errors.go", line:12} ("instances not found")
00:24:27 ... error stack:
00:24:27 	github.com/juju/juju/environs/errors.go:12: instances not found
00:24:27 
00:24:27 
00:24:27 ----------------------------------------------------------------------
00:24:27 PASS: environ_test.go:182: environSuite.TestPrecheck	0.000s
00:24:27 
00:24:27 ----------------------------------------------------------------------
00:24:27 FAIL: environ_test.go:211: controllerInstancesSuite.TestControllerInstances
00:24:27 
00:24:27 [LOG] 0:00.000 WARNING juju.environs.config unknown config field "controller-uuid"
00:24:27 [LOG] 0:00.000 WARNING juju.environs.config unknown config field "ca-cert"
00:24:27 [LOG] 0:00.000 WARNING juju.environs.config unknown config field "ca-private-key"
00:24:27 [LOG] 0:00.000 WARNING juju.environs.config unknown config field "controller-uuid"
00:24:27 [LOG] 0:00.000 WARNING juju.environs.config unknown config field "ca-cert"
00:24:27 [LOG] 0:00.000 WARNING juju.environs.config unknown config field "ca-private-key"
00:24:27 test 0
00:24:27 environ_test.go:244:
00:24:27     c.Assert(instances, gc.DeepEquals, []instance.Id{BootstrapInstanceId})
00:24:27 ... obtained []instance.Id = []instance.Id{"manual:hostname"}
00:24:27 ... expected []instance.Id = []instance.Id{"manual:"}
00:24:27 ... Difference:
00:24:27 ...     [0]: "manual:hostname" != "manual:"
00:24:27 
00:24:27 
00:24:27 
00:24:27 ----------------------------------------------------------------------
00:24:27 PASS: environ_test.go:259: controllerInstancesSuite.TestControllerInstancesError	0.002s
00:24:27 
00:24:27 ----------------------------------------------------------------------
00:24:27 FAIL: environ_test.go:266: controllerInstancesSuite.TestControllerInstancesInternal
00:24:27 
00:24:27 [LOG] 0:00.000 WARNING juju.environs.config unknown config field "ca-cert"
00:24:27 [LOG] 0:00.000 WARNING juju.environs.config unknown config field "ca-private-key"
00:24:27 [LOG] 0:00.000 WARNING juju.environs.config unknown config field "controller-uuid"
00:24:27 [LOG] 0:00.000 WARNING juju.environs.config unknown config field "ca-cert"
00:24:27 [LOG] 0:00.000 WARNING juju.environs.config unknown config field "ca-private-key"
00:24:27 [LOG] 0:00.000 WARNING juju.environs.config unknown config field "controller-uuid"
00:24:27 environ_test.go:274:
00:24:27     c.Assert(instances, gc.DeepEquals, []instance.Id{BootstrapInstanceId})
00:24:27 ... obtained []instance.Id = []instance.Id{"manual:hostname"}
00:24:27 ... expected []instance.Id = []instance.Id{"manual:"}
00:24:27 ... Difference:
00:24:27 ...     [0]: "manual:hostname" != "manual:"
00:24:27 
```

File: apiserver/facades/controller/migrationtarget/migrationtarget.go [hunk 1]
```diff
@@ -263,11 +264,22 @@ func (api *API) CheckMachines(args params.ModelArgs) (params.ErrorResults, error
...
+	// Begin check for manual cloud - we do not want to perform certain checks if the cloud
+	// is manaul
+	var model credentialcommon.Model
+	if model, err = backend.Model(); err != nil {
+		return params.ErrorResults{}, errors.Trace((err))
+	}
+	var cloud cloud.Cloud
+	if cloud, err = backend.Cloud(model.CloudName()); err != nil {
+		return params.ErrorResults{}, errors.Trace((err))
+	}
```
Comment (wallyworld) [inline apiserver/facades/controller/migrationtarget/migrationtarget.go:277]: you can just get the cloud like this, is more typical of what we do elsewhere
```
	model, err := st.Model();
	if err != nil {
		return params.ErrorResults{}, errors.Trace(err)
	}
	cloud, err := model.Cloud()
	if err != nil {
		return params.ErrorResults{}, errors.Trace(err)
	}
	return credentialcommon.ValidateExistingModelCredential(
		credentialcommon.NewPersistentBackend(st.State),
		context.CallContext(st.State),
		cloud.Type != "manual",
	)
```

File: state/cloudcredentials.go [hunk 2]
```diff
@@ -386,7 +386,7 @@ func validateCloudCredentials(
 }
 
 func validateCredentialForCloud(nuage cloud.Cloud, tag names.CloudCredentialTag, credential Credential) error {
-	if tag.Cloud().Id() != nuage.Name {
+	if tag.Cloud().Id() != nuage.Name && nuage.Name != "manual" {
```
Comment (wallyworld) [inline state/cloudcredentials.go:389]: Don't we want both the source cloud and target cloud to be "manual"?

File: provider/manual/environ.go [hunk 3]
```diff
@@ -74,7 +74,7 @@ func (*manualEnviron) StopInstances(context.ProviderCallContext, ...instance.Id)
 
 // AllInstances implements environs.InstanceBroker.
 func (e *manualEnviron) AllInstances(ctx context.ProviderCallContext) ([]instances.Instance, error) {
-	return e.Instances(ctx, []instance.Id{BootstrapInstanceId})
+	return e.Instances(ctx, []instance.Id{instance.Id(fmt.Sprintf("%s%s", BootstrapInstanceId, e.host))})
```
Comment (wallyworld) [inline provider/manual/environ.go:77]: Why this change? It's not mentioned in the PR and break unit tests.

**Review by wallyworld — COMMENTED** [review]
Core change to avoid instance checks looks ok (with suggested tweak). Ideally there'd be a unit test for the change.
But I'm unsure why we need the other changes?

File: apiserver/facades/controller/migrationtarget/migrationtarget.go (hunk 1 above)
Comment (VariableDeclared) [reply apiserver/facades/controller/migrationtarget/migrationtarget.go:277]: thanks!

File: provider/manual/environ.go (hunk 3 above)
Comment (VariableDeclared) [reply provider/manual/environ.go:77]: this was a drive-by that resolved "manual:" machine instance IDs, but I can pull this change our if you think it is appropriate

File: state/cloudcredentials.go (hunk 2 above)
Comment (VariableDeclared) [reply state/cloudcredentials.go:389]: you're right, this change was a left-over and should have been removed.

**Comment by VariableDeclared** [comment]
thanks for the review @wallyworld I think the changes are ready to your eyes again :)

File: provider/manual/environ.go (hunk 3 above)
Comment (VariableDeclared) [reply provider/manual/environ.go]: removed this

**Comment by VariableDeclared** [comment]
sure, thanks @wallyworld

**Comment by wallyworld** [comment]
Replaced by #13733
